### PR TITLE
Feature/winter 21  add service participants

### PR DESCRIPTION
**Pull request by bmuniswamy**
# Critical Changes

# Changes
  This is participant selection screen. This screen will display all the contacts associated with the program engagements for the given service. The users can select the participants that will be part of the sessions being created.
# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
~~- [ ] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))~~
~~- [ ] Base axe-core JEST test included for any new LWC (No critical violations)~~
~~- [ ] CRUD/FLS is enforced in Apex~~
~~- [ ] Update & Test permission sets to account for new features and CRUD/FLS~~
- [x] All modified classes are unit tested (Apex) at 100% coverage
- [x] UX approval or UX not necessary
- [x] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bmuniswamy** [comment]
[W-7981519](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008XFKQIA4/view)

**Comment by bmuniswamy** [comment]
@AndersonTarren , I had to add 3 labels for this work item. Sorry I missed it last week.

File: force-app/main/default/classes/SelectParticipantService.cls [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            Schema.SObjectType.ServiceParticipant__c.getName() => Schema.SObjectType.ServiceParticipant__c.getLabel(),
+            Schema.SObjectType.Contact.Fields.Name.getName() => Schema.SObjectType.Contact.Fields.Name.getLabel(),
+            Schema.SObjectType.Contact.Fields.Email.getName() => Schema.SObjectType.Contact.Fields.Email.getLabel(),
+            Schema.SObjectType.ProgramEngagement__c.Fields.Stage__c.getName() => Schema.SObjectType.ProgramEngagement__c.Fields.Stage__c.getLabel()
+        };
+
+        return objectFieldLabels;
+    }
+
+    public InitialDataAndSetup getInitialDataSetup(Id serviceId) {
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantService.cls:18]: `getSelectParticipantModel`

File: force-app/main/default/classes/SelectParticipantService.cls [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        dataSetup.programEngagements = selectParticipantSelector.getProgramEngagements(
+            serviceId
+        );
+        dataSetup.programCohorts = selectParticipantSelector.getProgramCohorts(serviceId);
+        dataSetup.labels = getObjectData();
+        return dataSetup;
+    }
+
+    public class InitialDataAndSetup {
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantService.cls:31]: `SelectParticipantModel`

File: force-app/main/default/classes/SelectParticipantService.cls [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        @AuraEnabled
+        public List<Object> programEngagements;
+        @AuraEnabled
+        public List<Object> programCohorts;
+        @AuraEnabled
+        public Map<String, String> labels;
+
+        public InitialDataAndSetup() {
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantService.cls:39]:
```suggestion
        public SelectParticipantModel() {
```

File: force-app/main/default/classes/SelectParticipantController.cls [hunk 4]
```diff
@@ -0,0 +1,15 @@
+public with sharing class SelectParticipantController {
+    @TestVisible
+    private static SelectParticipantService selectParticipantService = new SelectParticipantService();
+
+    @AuraEnabled(cacheable=true)
+    public static selectParticipantService.InitialDataAndSetup getInitialDataSetup(
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantController.cls:6]: capital S on `SelectParticipantService.SelectParticipantModel`

Comment (bethbrains) [reply force-app/main/default/classes/SelectParticipantController.cls:6]: `getSelectParticipantModel`

File: force-app/main/default/classes/SelectParticipantController_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,60 @@
+@isTest
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantController_TEST.cls:1]:
```suggestion
@IsTest
```

File: force-app/main/default/classes/SelectParticipantController_TEST.cls [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+public with sharing class SelectParticipantController_TEST {
+    private static BasicStub serviceStub = new BasicStub(SelectParticipantService.class);
+
+    @isTest
+    private static void testGetInitialDataSetup() {
+        final Id serviceId = TestUtil.mockId(Service__c.SObjectType);
+        final SelectParticipantService.InitialDataAndSetup expected = new SelectParticipantService.InitialDataAndSetup();
+
+        serviceStub.withReturnValue('getInitialDataSetup', Id.class, expected);
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantController_TEST.cls:10]: Use var for method name

File: force-app/main/default/classes/SelectParticipantSelector.cls [hunk 7]
```diff
@@ -0,0 +1,45 @@
+public with sharing class SelectParticipantSelector {
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantSelector.cls:1]:
```suggestion
public with sharing class ServiceParticipantSelector {
```

File: force-app/main/default/classes/SelectParticipantService.cls [hunk 8]
```diff
@@ -0,0 +1,45 @@
+public with sharing class SelectParticipantService {
+    @TestVisible
+    private SelectParticipantSelector selectParticipantSelector = new SelectParticipantSelector();
+
+    public Map<String, String> getObjectData() {
```
Comment (bethbrains) [inline force-app/main/default/classes/SelectParticipantService.cls:5]: getLabels

File: force-app/main/default/lwc/selectedParticipantList/selectedParticipantList.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+        let index = tempselectedParticipants.findIndex(
+            element => element.Id === event.detail.row.Id
+        );
+
+        tempselectedParticipants.splice(index, 1);
+        this.selectedParticipants = tempselectedParticipants;
+
+        this.noRecordsSelected =
+            this.selectedParticipants && this.selectedParticipants.length === 0;
```
Comment (bethbrains) [inline force-app/main/default/lwc/selectedParticipantList/selectedParticipantList.js:77]:
```suggestion
    handleRemove(event) {
```
The parent handles the actual removal from the array and passes back down the updated array.

**Review by bethbrains — CHANGES_REQUESTED** [review]
GREAT work. Great mob review.

**Review by jjbennett — COMMENTED** [review]
A couple nit picky things on the apex side, let's hop on a call for the js side of things.

**Review by jjbennett — APPROVED** [review]
A couple comments but no show stoppers, I know you still have the css to tweak on the right table but overall I think this is an amazing first pass. Thanks again for consolidating the code, I think it streamlines the communication!

**Comment by bmuniswamy** [comment]
enqueLoadrecords calls the applyFliter method that will be used when a user types in the input text search.

**Comment by bmuniswamy** [comment]
@jjbennett , I was setting the result to undefined if there is an error. I learnt that from the superbadge and if that is not the right approach please let me know

File: force-app/main/default/lwc/participantSelector/participantSelector.js [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+            this.engagements = result.data[ENGAGEMENTS].slice(0);
+            this.cohorts = result.data[COHORTS].slice(0);
+            let labels = result.data[LABELS];
+
+            this.loadDataTable();
+            this.loadProgramCohorts(this.cohorts);
+            this.loadLabels(labels);
+        } else if (error) {
+            console.log(error);
+            result = undefined;
```
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/lwc/participantSelector/participantSelector.js:148]: I believe the idea is to set the variables that held the results to undefined but not the result itself.

**Pull request merged**